src/crypto/secretstream/secretstream_macros.rs

```diff
@@ -169,19 +169,22 @@ impl Stream<Push> {
     /// encryption applications can use the password hashing API to get a key
     /// that can be used with the functions below.
     pub fn init_push(key: &Key) -> Result<(Stream<Push>, Header), ()> {
-        let mut header: [u8; HEADERBYTES] = unsafe { mem::uninitialized() };
-        let mut state: $state_name =
-            unsafe { mem::uninitialized() };
+        let mut header = mem::MaybeUninit::<[u8; HEADERBYTES]>::uninit();
+        let mut state = mem::MaybeUninit::uninit();
         let rc = unsafe {
             $init_push_name(
-                &mut state,
-                header.as_mut_ptr(),
+                state.as_mut_ptr(),
+                header.as_mut_ptr() as *mut u8,
```

Comment:
Are you sure about this one ? a double `as_mut_ptr()` should avoid the cast `header.as_mut_ptr().as_mut_ptr()`. And this make me think that `[mem::MaybeUninit::<u8>; HEADERBYTES]` should make more sense ?

Reply:
I feel like using `[mem::MaybeUninit::<u8>; HEADERBYTES]` would lead to uglier code. I'm ok with the pointer cast.

Reply:
Me too.

Reply:
Something like that would look better but currently require nightly:

```rust
let mut header: [mem::MaybeUninit::<u8>; HEADERBYTES] = unsafe { mem::MaybeUninit::uninit().assume_init() };
let mut state = mem::MaybeUninit::uninit();
let rc = unsafe {
    $init_push_name(
        state.as_mut_ptr(),
        mem::MaybeUninit::first_ptr_mut(&mut header),
        key.0.as_ptr(),
    )
};
if rc != 0 {
    return Err(());
}
// rc == 0 and both state and header are initialized
let state = unsafe { state.assume_init() };
let header = unsafe { mem::transmute(header) };
```

Reply:
If HEADERBYTES is small enough we can initialize header with zeroes.

> `unsafe { mem::transmute(header) };`

Please, no =)

Reply:
That what the doc of std advice https://doc.rust-lang.org/nightly/std/mem/union.MaybeUninit.html#initializing-an-array-element-by-element

Reply:
`[mem::MaybeUninit::<u8>::uninit(); HEADERBYTES]` should work to create an uninitialized array without `unsafe`; `MaybeUninit<u8>` is `Copy`.